File: python/breadth-first-search/WordLadder.py
from typing import List
from collections import deque

class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        word_set = set(wordList)
        
        queue = deque([beginWord])
        word_count = 1
        while queue:
            queue_size = len(queue)
            for i in range(queue_size):
                curr_word = queue.popleft()
                if curr_word == endWord:
                    return word_count
                
                word_list = list(curr_word)
                for j in range(len(word_list)):
                    original_char = word_list[j]
                    for c in 'abcdefghijklmnopqrstuvwxyz':
                        word_list[j] = c
                        next_word = ''.join(word_list)
                        if next_word in word_set:
                            queue.append(next_word)
                            word_set.remove(next_word)
                    word_list[j] = original_char
            
            word_count += 1
        
        return 0

File: java/breadth-first-search/WordLadder.java
import java.util.*;

class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> set = new HashSet<>(wordList);

        Queue<String> queue = new LinkedList<>();
        queue.offer(beginWord);

        int wordCount = 1;
        while(!queue.isEmpty()) {
            int queueSize = queue.size();
            for(int i = 0; i < queueSize; i++) {
                String currWord = queue.poll();
                if(currWord.equals(endWord)) return wordCount;
                char word[] = currWord.toCharArray();
                for(int j = 0; j < word.length; j++) {
                    char ch = word[j];
                    for(char c = 'a'; c <= 'z'; c++) {
                        word[j] = c;
                        String nextWord = new String(word);
                        if(set.contains(nextWord)) {
                            queue.add(nextWord);
                            set.remove(nextWord);
                        }
                    }
                    word[j] = ch;
                }
            }
            wordCount += 1;
        }
        return 0;
    }
}